From b909ec5ef765da27b14e25bd3094bb5214977974 Mon Sep 17 00:00:00 2001
From: Prasad Mujumdar <prasadm@cloudera.com>
Date: Wed, 14 Aug 2013 15:17:38 -0700
Subject: [PATCH 196/218] CLOUDERA-BUILD CDH-13569: Malformed hive-default.xml.template in nightly tarball

---
 conf/hive-default.xml.template |    4 +++-
 1 files changed, 3 insertions(+), 1 deletions(-)

diff --git a/src/conf/hive-default.xml.template b/src/conf/hive-default.xml.template
index a0bb3da..fbf6ca8 100644
--- a/src/conf/hive-default.xml.template
+++ b/src/conf/hive-default.xml.template
@@ -479,6 +479,7 @@
     If the user has set hive.merge.mapfiles to true and hive.merge.mapredfiles to false, the idea was the
     number of reducers are few, so the number of files anyway are small. However, with this optimization,
     we are increasing the number of files possibly by a big margin. So, we merge aggresively.
+  </description>
 </property>
 
 <property>
@@ -487,6 +488,7 @@
   <description>Whether the version of hadoop which is running supports sub-directories for tables/partitions.
     Many hive optimizations can be applied if the hadoop version supports sub-directories for
     tables/partitions. It was added by MAPREDUCE-1501
+  </description>
 </property>
 
 <property>
@@ -1693,7 +1695,7 @@
 
 <property>
   <name>hive.server2.thrift.sasl.qop</name>
-  <value>auth</auth>
+  <value>auth</value>
   <description>Sasl QOP value; one of 'auth', 'auth-int' and 'auth-conf'</description>
 </property>
 
-- 
1.7.0.4

