From 8cbe7c4b796cae0fc0e8fd8d0d511129d2e02515 Mon Sep 17 00:00:00 2001
From: stack <stack@apache.org>
Date: Thu, 7 Jan 2016 15:35:59 -0800
Subject: [PATCH 012/197] HBASE-15079 TestMultiParallel.validateLoadedData AssertionError: null

---
 .../hadoop/hbase/client/TestMultiParallel.java     |   39 ++++++++++++++++++--
 1 files changed, 35 insertions(+), 4 deletions(-)

diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestMultiParallel.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestMultiParallel.java
index 0469bca..c764129 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestMultiParallel.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/client/TestMultiParallel.java
@@ -669,13 +669,44 @@ public class TestMultiParallel {
   private void validateLoadedData(Table table) throws IOException {
     // get the data back and validate that it is correct
     LOG.info("Validating data on " + table);
+    List<Get> gets = new ArrayList<Get>();
     for (byte[] k : KEYS) {
       Get get = new Get(k);
       get.addColumn(BYTES_FAMILY, QUALIFIER);
-      Result r = table.get(get);
-      Assert.assertTrue(r.containsColumn(BYTES_FAMILY, QUALIFIER));
-      Assert.assertEquals(0, Bytes.compareTo(VALUE, r
-          .getValue(BYTES_FAMILY, QUALIFIER)));
+      gets.add(get);
+    }
+    int retryNum = 10;
+    Result[] results = null;
+    do  {
+      results = table.get(gets);
+      boolean finished = true;
+      for (Result result : results) {
+        if (result.isEmpty()) {
+          finished = false;
+          break;
+        }
+      }
+      if (finished) {
+        break;
+      }
+      try {
+        Thread.sleep(10);
+      } catch (InterruptedException e) {
+      }
+      retryNum--;
+    } while (retryNum > 0);
+
+    if (retryNum == 0) {
+      fail("Timeout for validate data");
+    } else {
+      if (results != null) {
+        for (Result r : results) {
+          Assert.assertTrue(r.containsColumn(BYTES_FAMILY, QUALIFIER));
+          Assert.assertEquals(0, Bytes.compareTo(VALUE, r
+            .getValue(BYTES_FAMILY, QUALIFIER)));
+        }
+        LOG.info("Validating data on " + table + " successfully!");
+      }
     }
   }
 
-- 
1.7.0.4

