From 8a36c2bffecdab35af725c8b725be6f4477d6cf7 Mon Sep 17 00:00:00 2001
From: tedyu <yuzhihong@gmail.com>
Date: Fri, 8 Jan 2016 15:59:08 -0800
Subject: [PATCH 018/197] HBASE-15057 local-master-backup.sh doesn't start HMaster correctly (Samir Ahmic)

---
 bin/local-master-backup.sh |    1 +
 1 files changed, 1 insertions(+), 0 deletions(-)

diff --git a/bin/local-master-backup.sh b/bin/local-master-backup.sh
index bd265ab..cc2dc56 100755
--- a/bin/local-master-backup.sh
+++ b/bin/local-master-backup.sh
@@ -40,6 +40,7 @@ run_master () {
   DN=$2
   export HBASE_IDENT_STRING="$USER-$DN"
   HBASE_MASTER_ARGS="\
+    -D hbase.master.port=`expr 16000 + $DN` \
     -D hbase.master.info.port=`expr 16010 + $DN` \
     -D hbase.regionserver.port=`expr 16020 + $DN` \
     -D hbase.regionserver.info.port=`expr 16030 + $DN` \
-- 
1.7.0.4

