From 36d1adb6065a20208ea2c43130abd021856011dd Mon Sep 17 00:00:00 2001
From: stack <stack@apache.org>
Date: Sat, 16 Jan 2016 15:53:07 -0800
Subject: [PATCH 032/197] HBASE-14962 TestSplitWalDataLoss fails on all branches

---
 .../hbase/regionserver/TestSplitWalDataLoss.java   |   11 ++++++++++-
 1 files changed, 10 insertions(+), 1 deletions(-)

diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSplitWalDataLoss.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSplitWalDataLoss.java
index 102e7bc..962412f 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSplitWalDataLoss.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/regionserver/TestSplitWalDataLoss.java
@@ -24,6 +24,7 @@ import static org.mockito.Mockito.spy;
 
 import java.io.IOException;
 import java.util.Collection;
+import java.util.Map;
 
 import org.apache.commons.lang.mutable.MutableBoolean;
 import org.apache.commons.logging.Log;
@@ -117,7 +118,15 @@ public class TestSplitWalDataLoss {
     }).when(spiedRegion).internalFlushCacheAndCommit(Matchers.<WAL> any(),
       Matchers.<MonitoredTask> any(), Matchers.<PrepareFlushResult> any(),
       Matchers.<Collection<Store>> any());
-    rs.onlineRegions.put(rs.onlineRegions.keySet().iterator().next(), spiedRegion);
+    // Find region key; don't pick up key for hbase:meta by mistake.
+    String key = null;
+    for (Map.Entry<String, Region> entry: rs.onlineRegions.entrySet()) {
+      if (entry.getValue().getRegionInfo().getTable().equals(this.tableName)) {
+        key = entry.getKey();
+        break;
+      }
+    }
+    rs.onlineRegions.put(key, spiedRegion);
     Connection conn = testUtil.getConnection();
 
     try (Table table = conn.getTable(tableName)) {
-- 
1.7.0.4

