From 7762b66da3f29f1e41aaec3a37d451f6111a58b8 Mon Sep 17 00:00:00 2001
From: stack <stack@apache.org>
Date: Wed, 20 Jan 2016 09:45:49 -0800
Subject: [PATCH 042/197]   HBASE-15098 Normalizer switch in configuration is not used; addendum

---
 .../TestSimpleRegionNormalizerOnCluster.java       |    1 -
 1 files changed, 0 insertions(+), 1 deletions(-)

diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/master/normalizer/TestSimpleRegionNormalizerOnCluster.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/master/normalizer/TestSimpleRegionNormalizerOnCluster.java
index 0bab6aa..6b23240 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/master/normalizer/TestSimpleRegionNormalizerOnCluster.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/master/normalizer/TestSimpleRegionNormalizerOnCluster.java
@@ -61,7 +61,6 @@ public class TestSimpleRegionNormalizerOnCluster {
   public static void beforeAllTests() throws Exception {
     // we will retry operations when PleaseHoldException is thrown
     TEST_UTIL.getConfiguration().setInt(HConstants.HBASE_CLIENT_RETRIES_NUMBER, 3);
-    TEST_UTIL.getConfiguration().setBoolean(QuotaUtil.QUOTA_CONF_KEY, true);
 
     // Start a cluster of two regionservers.
     TEST_UTIL.startMiniCluster(1);
-- 
1.7.0.4

