From a852cba5dfdf67aa69ba0bbe02b45a58995e96a9 Mon Sep 17 00:00:00 2001
From: Mikhail Antonov <antonov@apache.org>
Date: Wed, 20 Jan 2016 13:50:32 -0800
Subject: [PATCH 044/197] HBASE-15139 Connection manager doesn't pass client metrics to RpcClient

---
 .../hadoop/hbase/client/ConnectionManager.java     |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionManager.java b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionManager.java
index 2efa1af..24f7c60 100644
--- a/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionManager.java
+++ b/hbase-client/src/main/java/org/apache/hadoop/hbase/client/ConnectionManager.java
@@ -650,7 +650,7 @@ class ConnectionManager {
       this.registry = setupRegistry();
       retrieveClusterId();
 
-      this.rpcClient = RpcClientFactory.createClient(this.conf, this.clusterId);
+      this.rpcClient = RpcClientFactory.createClient(this.conf, this.clusterId, this.metrics);
       this.rpcControllerFactory = RpcControllerFactory.instantiate(conf);
 
       // Do we publish the status?
-- 
1.7.0.4

