From 936203684fe2c4751d9e4d3aaf232bed1d896198 Mon Sep 17 00:00:00 2001
From: tedyu <yuzhihong@gmail.com>
Date: Fri, 22 Jan 2016 02:42:10 -0800
Subject: [PATCH 048/197] HBASE-15133 Data loss after compaction when a row has more than Integer.MAX_VALUE columns (Toshihiro Suzuki)

---
 .../hadoop/hbase/regionserver/StoreScanner.java    |    2 +-
 1 files changed, 1 insertions(+), 1 deletions(-)

diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreScanner.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreScanner.java
index 24042b7..a026a40 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreScanner.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/regionserver/StoreScanner.java
@@ -64,7 +64,7 @@ public class StoreScanner extends NonReversedNonLazyKeyValueScanner
   protected KeyValueHeap heap;
   protected boolean cacheBlocks;
 
-  protected int countPerRow = 0;
+  protected long countPerRow = 0;
   protected int storeLimit = -1;
   protected int storeOffset = 0;
 
-- 
1.7.0.4

