From cd920940b736f1674780a4f8f5fd888023a3551e Mon Sep 17 00:00:00 2001
From: tedyu <yuzhihong@gmail.com>
Date: Fri, 22 Jan 2016 02:50:02 -0800
Subject: [PATCH 049/197] HBASE-15148 Resolve IS2_INCONSISTENT_SYNC findbugs warning in AuthenticationTokenSecretManager (Yu Li)

---
 .../token/AuthenticationTokenSecretManager.java    |    9 +++++----
 1 files changed, 5 insertions(+), 4 deletions(-)

diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/security/token/AuthenticationTokenSecretManager.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/security/token/AuthenticationTokenSecretManager.java
index 09b8846..0951093 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/security/token/AuthenticationTokenSecretManager.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/security/token/AuthenticationTokenSecretManager.java
@@ -287,6 +287,10 @@ public class AuthenticationTokenSecretManager
     }
   }
 
+  synchronized long getLastKeyUpdate() {
+    return lastKeyUpdate;
+  }
+
   public static SecretKey createSecretKey(byte[] raw) {
     return SecretManager.createSecretKey(raw);
   }
@@ -340,10 +344,7 @@ public class AuthenticationTokenSecretManager
 
         // clear any expired
         removeExpiredKeys();
-        long localLastKeyUpdate;
-        synchronized (this) {
-          localLastKeyUpdate = lastKeyUpdate;
-        }
+        long localLastKeyUpdate = getLastKeyUpdate();
         if (localLastKeyUpdate + keyUpdateInterval < now) {
           // roll a new master key
           rollCurrentKey();
-- 
1.7.0.4

