From 5cae3f66a910042a564e741d74007253ad2ab10b Mon Sep 17 00:00:00 2001
From: stack <stack@apache.org>
Date: Sun, 31 Jan 2016 10:51:41 -0600
Subject: [PATCH 056/197] HBASE-15190 Monkey dies when running on shared cluster (gives up when can't kill the other fellows processes)

---
 .../apache/hadoop/hbase/HBaseClusterManager.java   |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/hbase-it/src/test/java/org/apache/hadoop/hbase/HBaseClusterManager.java b/hbase-it/src/test/java/org/apache/hadoop/hbase/HBaseClusterManager.java
index c49ae44..ba6a4a9 100644
--- a/hbase-it/src/test/java/org/apache/hadoop/hbase/HBaseClusterManager.java
+++ b/hbase-it/src/test/java/org/apache/hadoop/hbase/HBaseClusterManager.java
@@ -174,7 +174,7 @@ public class HBaseClusterManager extends Configured implements ClusterManager {
     }
 
     protected String findPidCommand(ServiceType service) {
-      return String.format("ps aux | grep proc_%s | grep -v grep | tr -s ' ' | cut -d ' ' -f2",
+      return String.format("ps ux | grep proc_%s | grep -v grep | tr -s ' ' | cut -d ' ' -f2",
           service);
     }
 
@@ -269,7 +269,7 @@ public class HBaseClusterManager extends Configured implements ClusterManager {
 
     @Override
     protected String findPidCommand(ServiceType service) {
-      return String.format("ps aux | grep %s | grep -v grep | tr -s ' ' | cut -d ' ' -f2",
+      return String.format("ps ux | grep %s | grep -v grep | tr -s ' ' | cut -d ' ' -f2",
         service);
     }
   }
-- 
1.7.0.4

