From c746f1250d90b4386ec3d4db035fa741e845d093 Mon Sep 17 00:00:00 2001
From: Apekshit <apeksharma@gmail.com>
Date: Wed, 3 Feb 2016 16:05:37 -0800
Subject: [PATCH 068/197] HBASE-15209 (compatibility) Disable table in HBaseTestingUtility.truncateTable. (Apekshit)

Signed-off-by: stack <stack@apache.org>
---
 .../apache/hadoop/hbase/HBaseTestingUtility.java   |    3 +++
 1 files changed, 3 insertions(+), 0 deletions(-)

diff --git a/hbase-server/src/test/java/org/apache/hadoop/hbase/HBaseTestingUtility.java b/hbase-server/src/test/java/org/apache/hadoop/hbase/HBaseTestingUtility.java
index a806127..e92e44d 100644
--- a/hbase-server/src/test/java/org/apache/hadoop/hbase/HBaseTestingUtility.java
+++ b/hbase-server/src/test/java/org/apache/hadoop/hbase/HBaseTestingUtility.java
@@ -2077,6 +2077,9 @@ public class HBaseTestingUtility extends HBaseCommonTestingUtility {
   public HTable truncateTable(final TableName tableName, final boolean preserveRegions)
       throws IOException {
     Admin admin = getHBaseAdmin();
+    if (!admin.isTableDisabled(tableName)) {
+      admin.disableTable(tableName);
+    }
     admin.truncateTable(tableName, preserveRegions);
     return new HTable(getConfiguration(), tableName);
   }
-- 
1.7.0.4

