From 63adcf33bf912aafe40a714408e6fd76d058c49b Mon Sep 17 00:00:00 2001
From: tedyu <yuzhihong@gmail.com>
Date: Fri, 5 Feb 2016 20:54:54 -0800
Subject: [PATCH 069/197] HBASE-15220 Change two logs in SimpleRegionNormalizer to INFO level

---
 .../master/normalizer/SimpleRegionNormalizer.java  |    4 ++--
 1 files changed, 2 insertions(+), 2 deletions(-)

diff --git a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/normalizer/SimpleRegionNormalizer.java b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/normalizer/SimpleRegionNormalizer.java
index ef9c102..56032d6 100644
--- a/hbase-server/src/main/java/org/apache/hadoop/hbase/master/normalizer/SimpleRegionNormalizer.java
+++ b/hbase-server/src/main/java/org/apache/hadoop/hbase/master/normalizer/SimpleRegionNormalizer.java
@@ -133,7 +133,7 @@ public class SimpleRegionNormalizer implements RegionNormalizer {
       // if the region is > 2 times larger than average, we split it, split
       // is more high priority normalization action than merge.
       if (regionSize > 2 * avgRegionSize) {
-        LOG.debug("Table " + table + ", large region " + hri.getRegionNameAsString() + " has size "
+        LOG.info("Table " + table + ", large region " + hri.getRegionNameAsString() + " has size "
             + regionSize + ", more than twice avg size, splitting");
         plans.add(new SplitNormalizationPlan(hri, null));
       } else {
@@ -143,7 +143,7 @@ public class SimpleRegionNormalizer implements RegionNormalizer {
         HRegionInfo hri2 = tableRegions.get(candidateIdx+1);
         long regionSize2 = getRegionSize(hri2);
         if (regionSize + regionSize2 < avgRegionSize) {
-          LOG.debug("Table " + table + ", small region size: " + regionSize
+          LOG.info("Table " + table + ", small region size: " + regionSize
             + " plus its neighbor size: " + regionSize2
             + ", less than the avg size " + avgRegionSize + ", merging them");
           plans.add(new MergeNormalizationPlan(hri, hri2));
-- 
1.7.0.4

