From 146c5285030aab67ac172b659e8b748335cd2275
From: Martin Kleppmann <martinkl@apache.org>
Date:   Wed Dec 30 15:43:43 2015 +0000
Subject: [PATCH] AVRO-1617: C: Fix equality checking of record schemas. Contributed by Skye Wanderman-Milne.

diff --git a/thirdparty/avro-c-1.7.4/src/schema_equal.c b/thirdparty/avro-c-1.7.4/src/schema_equal.c
index 287d218..0be1e83 100644
--- a/thirdparty/avro-c-1.7.4/src/schema_equal.c
+++ b/thirdparty/avro-c-1.7.4/src/schema_equal.c
@@ -33,6 +33,10 @@ schema_record_equal(struct avro_record_schema_t *a,
 	if (nullstrcmp(a->space, b->space)) {
 		return 0;
 	}
+	if (a->fields->num_entries != b->fields->num_entries) {
+		/* They have different numbers of fields */
+		return 0;
+	}
 	for (i = 0; i < a->fields->num_entries; i++) {
 		union {
 			st_data_t data;
