diff --git a/thirdparty/re2/re2/re2.cc b/thirdparty/re2/re2/re2.cc
index 8d1d468..00fad62 100644
--- a/thirdparty/re2/re2/re2.cc
+++ b/thirdparty/re2/re2/re2.cc
@@ -44,6 +44,7 @@ RE2::Options::Options(RE2::CannedOptions opt)
     max_mem_(kDefaultMaxMem),
     literal_(false),
     never_nl_(false),
+    dot_nl_(false),
     never_capture_(false),
     case_sensitive_(true),
     perl_classes_(false),
@@ -151,6 +152,9 @@ int RE2::Options::ParseFlags() const {
   if (never_nl())
     flags |= Regexp::NeverNL;
 
+  if (dot_nl())
+    flags |= Regexp::DotNL;
+
   if (never_capture())
     flags |= Regexp::NeverCapture;
 
diff --git a/thirdparty/re2/re2/re2.h b/thirdparty/re2/re2/re2.h
index 272028b..61ed484 100644
--- a/thirdparty/re2/re2/re2.h
+++ b/thirdparty/re2/re2/re2.h
@@ -512,6 +512,7 @@ class RE2 {
     //   max_mem          (see below)  approx. max memory footprint of RE2
     //   literal          (false) interpret string as literal, not regexp
     //   never_nl         (false) never match \n, even if it is in regexp
+    //   dot_nl           (false) dot matches everything including newline
     //   never_capture    (false) parse all parens as non-capturing
     //   case_sensitive   (true)  match is case-sensitive (regexp can override
     //                              with (?i) unless in posix_syntax mode)
@@ -567,6 +568,7 @@ class RE2 {
       max_mem_(kDefaultMaxMem),
       literal_(false),
       never_nl_(false),
+      dot_nl_(false),
       never_capture_(false),
       case_sensitive_(true),
       perl_classes_(false),
@@ -608,6 +610,9 @@ class RE2 {
     bool never_nl() const { return never_nl_; }
     void set_never_nl(bool b) { never_nl_ = b; }
 
+    bool dot_nl() const { return dot_nl_; }
+    void set_dot_nl(bool b) { dot_nl_ = b; }
+
     bool never_capture() const { return never_capture_; }
     void set_never_capture(bool b) { never_capture_ = b; }
 
@@ -631,6 +636,7 @@ class RE2 {
       max_mem_ = src.max_mem_;
       literal_ = src.literal_;
       never_nl_ = src.never_nl_;
+      dot_nl_ = src.dot_nl_;
       never_capture_ = src.never_capture_;
       case_sensitive_ = src.case_sensitive_;
       perl_classes_ = src.perl_classes_;
@@ -648,6 +654,7 @@ class RE2 {
     int64_t max_mem_;
     bool literal_;
     bool never_nl_;
+    bool dot_nl_;
     bool never_capture_;
     bool case_sensitive_;
     bool perl_classes_;
